Add new pagination interfaces.

ADmad opened the pull request:
Currently the pagination info is passed to the view using a request attribute. That's pretty hackish IMO and a result of legacy.

My proposed alternative is to have the paginator return an object which contains the results as well as the paging params/options. So `PaginatorInterface::paginate()` would return an instance of the new `Pager\PaginationInterface` which would be used a a view variable.

### The benefits of this approach are:
- This `PaginationInterface` instance can be iterated just like a resultset and also used by the `PaginationHelper`. The helper will no longer have to pluck info out of the request attribute.
- `PaginationInterface` has various helper methods to get specific params to avoid having to helper code be a maze of array keys.
- This opens the possibility to potentially create a serializable `PaginationInterface` instance for easily setting pagination info for API responses.

### Other changes:
- The new `Pager\PaginatorInterface` opens the possibility to potentially paginate any object like collections, arrays. (`Datasource\PaginatorInterface` could probably be removed).
- I have renamed some of the keys of paging params to be more intuitive.

### To do:
- [x] Update `Controller::paginate()` to no longer set the `paging` request attribute.
- [x] `PaginatorHelper` still needs to be updated to use `PaginationInterface` instances. Awaiting for the idea to be approved in principle before proceeding on that.
- [x] I am not set on the namespacing, interface names and methods for `PaginationInterface`. Am open to ideas if folks think existing ones could be improved.

ADmad committed:
Add new pagination interfaces.

The paginator now returns a PaginationInterface instance which holds
the paginated items as well as the paging params.

markstory commented on src/Pager/PaginationInterface.php:
```diff
@@ -0,0 +1,44 @@
+
+    public function hasNextPage(): bool;
+
+    public function items(): iterable;
```
Will users need to use `items()` when trying to get the results of pagination in their views?

ADmad replied:
No, the `PaginatedResult` class used by default extends `IteratorIterator`, so they can directly iterate it without having to use the `items()` method. There are already tests which iterate the result of `Controller::paginate()` which remain unchanged.

I have intentionally not made `PaginatedResultSet` implement the `CollectionInterface` because I think it's uncommon to do further operations on paginated results. Users always have the convenience of `collection()` if needed.

markstory commented on tests/TestCase/Datasource/SimplePaginatorTest.php:
```diff
@@ -55,39 +55,39 @@ public function testPaginateCustomFind(): void
         $pagingParams = $this->Paginator->getPagingParams();
-        $this->assertSame(3, $pagingParams['PaginatorPosts']['current']);
-        $this->assertNull($pagingParams['PaginatorPosts']['count']);
+        $this->assertSame(1, $pagingParams['PaginatorPosts']['currentPage']);
+        $this->assertNull($pagingParams['PaginatorPosts']['totalCount']);
```
I like the new names they are clearer. If we're adding new names could we also backport them to 4.x so that folks who rely on these keys have a chance to upgrade early?

ADmad replied:
IMO that would cause unnecessary grief to 4.x users who are directly dealing with the paging params. Array keys changes unlike API change are not easy to communicate. The `count` for e.g. now holds the resultset count instead of total count and can cause bugs in existing code.

There are community plugins too which use these keys:

https://github.com/FriendsOfCake/crud/blob/ace4262d8c1c6d4b26937cb19f8d3b1b954aeb4b/src/Listener/ApiPaginationListener.php#L55-L60

As a user I would rather upgrade these keys when I upgrade to next major instead of having to do that now.

markstory replied:
Is there a way we could backport the new implementation and have users opt-in to the new implementation using config options? I'm trying to figure out how we make the 5.x upgrade process like the 4.x one, where once you had fixed deprecation warnings you didn't have much else to do.

ADmad replied:
Not having to deal with the pain of trying to have 2 implementations together is why I did this for 5.x. For majority of the users these changes would be transparent. The only change they would have to do when upgrading is explicitly set the the paginated resultset to the paginator helper.

markstory commented on tests/TestCase/Datasource/SimplePaginatorTest.php:
```diff
@@ -108,19 +108,19 @@ public function testPaginateCustomFindFieldsArray(): void
         $results = $this->Paginator->paginate($table, [], $settings);
 
-        $result = $results->toArray();
+        $result = $results->items()->toArray();
```
Should the new object have a `toArray()` method on it?

ADmad replied:
I can add it.

ADmad committed:
Don't add paging params as request attribute.

ADmad committed:
Update PaginatorHelper to use PaginationInterface instance.

ADmad committed:
Remove unneeded methods

ADmad committed:
Remove unused argument

ADmad committed:
Code cleanup

ADmad commented:
I have updated `PaginatorHelper` to use the new `PaginationInterface` instances. 

Since paging data is no longer fetched from the request instance one needs to explicitly set the `PaginationInterface` to the helper using `PaginatorHelper::setPagination()` before using other methods. I hope this additional step won't be an issue. I could search the view vars and auto select the first `PaginationInterface` instance found as the default one but I would rather not add such "magic".

othercorey commented:
If we are introducing the `Pager` namespace here, I would suggest `Paging` as an alternative.

ADmad committed:
Change namespace Cake\Pager to Cake\Paging

ADmad commented:
Changed the namespace from `Pager` to `Paging`. Does `PaginationInterface` sound good (for the object which hold results and paging params) or should we change it to `PaginatedInterface`.

othercorey commented:
> Changed the namespace from `Pager` to `Paging`. Does `PaginationInterface` sound good (for the object which hold results and paging params) or should we change it to `PaginatedInterface`.

PaginatedInterface sounds like a result set. Is that what you want?

ADmad commented:
> PaginatedInterface sounds like a result set. Is that what you want?

Yup.

ADmad committed:
Rename PaginationInterface to PaginatedInterface.

markstory commented on src/Datasource/PaginatedResultSet.php:
```diff
@@ -0,0 +1,143 @@
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Datasource;
```
Should this class be in the `Paging` namespace with other paging related classes? Or should the new `Paging` namespace be `Datasource\Paging`?

othercorey replied:
The discontinuity is a little confusing to me as well. I can see maybe Paging was meant as an interface-only, but I'm not sure we need that kind of separation from the result set. I fear we're going to get into more Interface + Trait combinations to try and satisfy these split dependencies.

ADmad replied:
Initially I had thought of having `Paging` as a separate package. That would mean adding it as a dependency of the `Datasource` package.

But if the consensus is that having a separate paging package isn't necessary or will be useful enough then I can simplify things and move `Paging` to `Datasource/Paging` instead.

ADmad committed:
Move all paging related interfaces/classes under Datasource/Paging.

ADmad committed:
Auto search for PaginatedInterface.

If a PaginatedInterface instance is not explicity set for PaginatorHelper
then search the view vars and use the 1st one found.

ADmad commented:
To keep migration from 4.x simple I have removed the need to explicitly set the paginated resultset to the PaginatorHelper. For normal uses cases one would simply set the result of `Controller::paginate()` as a view var as it's currently done. Only if using scoped pagination one would have to set the correct resultset to the helper using `PaginatorHelper::setPaginated()`.

For cells when migration one would have to just drop the code which fetches the paging params using `Paginator::getParams()` and injects them in the request as `paging` attribute.

ADmad commented:
All paging related classes are now under `Datasource/Paging`. For some of the existing classes that I have moved, we can move them in 4.next itself and keep aliases for old locations as we normally do.

ADmad commented:
Any more concerns that need to be addressed here?

Earlier today/yesterday a user on Slack was having trouble getting the paging params from the request attribute as he was unaware when that attribute is set.

So I think this change is in the right direction with regards to making access to paging data more explicit.

ADmad committed:
Merge branch '5.x' into 5.0-pager
